/*
 Copyright 2012 Google Inc. All Rights Reserved.
 Author: blakely@google.com (Tim Blakely)

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

syntax = "proto2";

import "basics.proto";

package bigbrain;


// General state descriptions for neurons, channels, and intrinsics.  These are
// common to all Neurons/Channels.  As of now, they both have very common
// members, but I expect them to diverge soon.
message CommonParams {
  optional Position pos = 1;
  optional int32 random_seed = 2 [default = 0];
  optional string topo = 7;
  optional string name = 3;
  optional uint64 uuid = 6 [default = 0];
  optional DiffEqSolver diff_eq_solver = 5 [default = RUNGEKUTTA];
  optional Color color = 8;
}

message Connection {
  required uint64 from_uuid = 1;
  required uint64 to_uuid = 2;
}

//
// Potassium and sodium channel
//

// IKNa State
message IKNaState {
  optional double output = 1 [default = 0];
  optional double g = 2 [default = 0];
  optional double na = 3 [default = 0];
}


// IKNa Specific Parameters
message IKNaParams {
  optional double g_peak = 2 [default = 1];
  optional double e_rev = 3 [default = -90];
  optional double sodium_influx_peak = 4 [default = 0.025];
  optional double sodium_equilibrium = 5 [default = 0.001];
  optional double sodium_time_constant = 6 [default = 1250];
  optional double sodium_threshold = 7 [default = -10];
  optional double sodium_threshold_slope = 8 [default = 5];
}

//
// Persistent Sodium
//

message INaPState {
  optional double output = 1 [default = 0];
  optional double g = 2 [default = 0];
}

message INaPParams {
  optional double g_peak = 1 [default = 1];
  optional double e_rev = 2 [default = 30.0];
  optional double threshold = 3 [default = -55.7];
  optional double slope = 4 [default = 7.7];
}

//
// Cation
//

message IhState {
  optional double output = 3 [default = 0];
  optional double g = 1 [default = 0];
  optional double m = 2 [default = 0];
}

message IhParams {
  optional double g_peak = 1 [default = 1];
  optional double e_rev = 2 [default = -40];
  optional double v_threshold = 3 [default = -75];
}
//
// Pulse Integrate And Fire neuron
//

// PIAF state

message PulseIntegrateAndFireState {
  optional uint32 firing_state = 1 [default = 4];
  optional double membrane_potential = 2 [default = 0];
  optional double threshold = 3 [default = 0];
  optional double last_plastic_time = 4 [default = 0];
  optional double last_spike_time = 5 [default = 0];
}


message PulseIntegrateAndFireParams {
  optional double sodium_leak_conductance = 2 [default = 0.2];
  optional double potassium_leak_conductance = 3 [default = 1.0];
  optional double membrane_time_constant = 4 [default = 16.0];
  optional double threshold_time_constant = 5 [default = 2];
  optional double resting_threshold = 6 [default = -51];
  optional double potassium_reversal_potential = 7 [default = -90];
  optional double sodium_reversal_potential = 8 [default = 30];
  optional double spike_time_constant = 9 [default = 1.75];
  optional double spike_duration = 10 [default = 2];
  optional bool plasticity = 12 [default = false];
}

//
// AMPA Synapse
//

message AmpaState {
  optional double input = 1;
  optional double output = 2;
  optional double dg_dt = 3;
  optional double g = 4;
  optional double g_lower = 5;
}

message AmpaParams {
  optional double time_constant_1 = 1 [default = 0.5];
  optional double time_constant_2 = 2 [default = 2.4];
  optional double g_peak = 3 [default = 0.1];
  optional double reverse_potential = 4 [default = 0];
  optional double mean_mini_amplitude = 5 [default = 0.5];
  optional double std_mini_amplitude = 6 [default = 0.25];
  optional double noise_rate = 7 [default = 0];
  optional double noise_amplitude = 8 [default = 0];
  optional string noise_mode = 9 [default = "Poisson"];
  optional string conductance_table = 10 [default = "none"];
  optional string noise_cutoff_table = 11 [default = "none"];
  optional double peak_response = 15 [default = 0];
  optional double probability = 16 [default = 1.0];

  // Should be in some common synapse param set
  optional bool delay = 14 [default = false];
  optional double mean_delay = 12 [default = 0];
  optional double std_std = 13 [default = 0];

}
